File: lib/elixir/runner.ex
defmodule AocLib.Runner do
  @moduledoc """
  Runner for executing Advent of Code solutions.
  """

  alias AocLib.Solution
  alias AocLib.Types.{ProblemInput, ProblemResult}

  defstruct [:year, :solutions]

  @type t :: %__MODULE__{
          year: non_neg_integer(),
          solutions: [module() | nil]
        }

  @sample_str ["real", "samp"]

  @doc """
  Creates a new runner for a given year with a list of solution modules.
  """
  @spec new(non_neg_integer(), [module() | nil]) :: t()
  def new(year, solutions) do
    %__MODULE__{year: year, solutions: solutions}
  end

  @doc """
  Parses command line arguments and returns configuration.
  Returns {:ok, command, opts} or {:error, message} or :help
  """
  @spec parse_args([String.t()]) :: {:ok, String.t() | nil, map()} | {:error, String.t()} | :help
  def parse_args(args) do
    {opts, args, invalid} =
      OptionParser.parse(args,
        strict: [sample: :boolean, part: :integer, all: :boolean, help: :boolean],
        aliases: [s: :sample, p: :part, a: :all, h: :help]
      )

    cond do
      opts[:help] ->
        :help

      invalid != [] ->
        {:error, "Invalid options: #{inspect(invalid)}"}

      true ->
        command = List.first(args)
        validate_options(command, opts, args)
    end
  end

  defp validate_options(command, opts, args) do
    use_sample = Keyword.get(opts, :sample, false)
    part = Keyword.get(opts, :part)
    all = Keyword.get(opts, :all, false)

    case command do
      "main" ->
        # main ignores --part, --sample, --all
        {:ok, "main", %{all: true, part: 1, use_sample: false}}

      "day" ->
        day_num = Enum.at(args, 1)

        case day_num do
          nil ->
            {:error, "Day number required after 'day' command"}

          day_str ->
            case Integer.parse(day_str) do
              {day, ""} when day >= 1 and day <= 25 ->
                cond do
                  all ->
                    # --all is given, part is ignored
                    {:ok, "day#{String.pad_leading(Integer.to_string(day), 2, "0")}",
                     %{all: true, part: 1, use_sample: use_sample}}

                  part == nil ->
                    {:error, "--part <1|2> is required unless --all is given"}

                  part not in [1, 2] ->
                    {:error, "--part must be 1 or 2, got: #{part}"}

                  true ->
                    {:ok, "day#{String.pad_leading(Integer.to_string(day), 2, "0")}",
                     %{all: false, part: part, use_sample: use_sample}}
                end

              _ ->
                {:error, "Invalid day number: #{day_str}"}
            end
        end

      nil ->
        {:error, "No command specified. Use 'main', 'day N', or 'test'"}

      _ ->
        {:error, "Unknown command: #{command}. Use 'main', 'day N', or 'test'"}
    end
  end

  @doc """
  Runs the solutions based on parsed configuration.
  """
  @spec run(t(), String.t() | nil, boolean(), 1 | 2, boolean()) :: :ok
  def run(runner, arg, full_day, part, use_sample) do
    case arg do
      nil ->
        IO.puts("Usage: elixir main.exs <day>")

      "main" ->
        run_full_year(runner)

      "day" <> day_str ->
        case Integer.parse(day_str) do
          {day, ""} when day >= 1 and day <= 25 ->
            if full_day do
              run_day(runner, day)
            else
              run_single(runner, day, part, use_sample)
            end

          _ ->
            IO.puts("Argument should have format dayXX: #{arg}")
        end

      _ ->
        IO.puts("Argument should be 'main' or 'dayXX': #{arg}")
    end
  end

  @doc """
  Verifies that a solution matches the expected result.
  """
  @spec verify_solution(t(), non_neg_integer(), 1 | 2, boolean()) ::
          :ok | {:error, String.t()}
  def verify_solution(runner, day, part, use_sample) do
    solution = Enum.at(runner.solutions, day - 1)

    if solution == nil do
      {:error, "No solution implemented for day #{day}"}
    else
      expected_index = (part - 1) * 2 + if(use_sample, do: 0, else: 1)
      expected = Enum.at(solution.results(), expected_index)

      case get_result(runner, day, part, use_sample) do
        {:ok, {result, _duration}} ->
          if ProblemResult.equal?(result, expected) do
            :ok
          else
            {:error,
             "Got #{ProblemResult.format(result)}, expected #{ProblemResult.format(expected)}"}
          end

        {:error, reason} ->
          {:error, "Error: #{reason}"}
      end
    end
  end

  @doc """
  Verifies all solutions against expected results.
  Returns true if all tests pass.
  """
  @spec verify_solutions(t()) :: boolean()
  def verify_solutions(runner) do
    IO.puts("\n----------")

    success =
      runner.solutions
      |> Enum.with_index(1)
      |> Enum.reduce(true, fn {solution, day}, acc ->
        if solution != nil do
          # Check all 4 combinations for this day
          day_success =
            for part <- [1, 2], use_sample <- [true, false], reduce: true do
              day_acc ->
                sample_label = if use_sample, do: "s", else: "r"

                case get_result(runner, day, part, use_sample) do
                  {:ok, {result, _duration}} ->
                    expected_index = (part - 1) * 2 + if(use_sample, do: 0, else: 1)
                    expected = Enum.at(solution.results(), expected_index)

                    IO.write(
                      "Testing D#{String.pad_leading(Integer.to_string(day), 2, "0")} P#{part} '#{sample_label}': "
                    )

                    if ProblemResult.equal?(result, expected) do
                      IO.puts("PASSED")
                      day_acc
                    else
                      IO.puts(
                        "FAILED: Got #{ProblemResult.format(result)}, expected #{ProblemResult.format(expected)}"
                      )

                      false
                    end

                  {:error, :missing_solution} ->
                    day_acc

                  {:error, reason} ->
                    IO.write(
                      "Testing D#{String.pad_leading(Integer.to_string(day), 2, "0")} P#{part} '#{sample_label}': "
                    )

                    IO.puts("ERROR RESULT (#{reason})")
                    false
                end
            end

          acc and day_success
        else
          acc
        end
      end)

    IO.puts("----------")
    success
  end

  defp run_full_year(runner) do
    IO.puts("")

    {total_elapsed, _} =
      :timer.tc(fn ->
        runner.solutions
        |> Enum.with_index(1)
        |> Enum.each(fn {solution, day} ->
          if solution != nil do
            IO.puts("Day #{String.pad_leading(Integer.to_string(day), 2, "0")}:")

            day_elapsed =
              for part <- [1, 2],
                  use_sample <- [true, false],
                  reduce: 0.0 do
                acc ->
                  input = Enum.at(@sample_str, if(use_sample, do: 0, else: 1))

                  case get_result(runner, day, part, use_sample) do
                    {:ok, {result, elapsed}} ->
                      IO.puts("  P#{part} #{input}:  #{ProblemResult.format(result)}")
                      acc + elapsed

                    {:error, _reason} ->
                      IO.puts("  P#{part} #{input}:  <Error>")
                      acc
                  end
              end

            IO.puts("  > Runtime: #{:io_lib.format("~.4f", [day_elapsed])}s\n")
          end
        end)
      end)

    total_secs = total_elapsed / 1_000_000
    IO.puts("\n\nTotal Runtime: #{:io_lib.format("~.4f", [total_secs])}s")
  end

  defp run_day(runner, day) do
    solution = Enum.at(runner.solutions, day - 1)

    if solution == nil do
      IO.puts(
        "No solution implemented for day #{String.pad_leading(Integer.to_string(day), 2, "0")} in year #{runner.year}"
      )
    else
      expected_results = solution.results()

      {day_elapsed, matches_expected} =
        for part <- [1, 2], use_sample <- [true, false], reduce: {0.0, true} do
          {time_acc, match} ->
            case get_result(runner, day, part, use_sample) do
              {:ok, {result, elapsed}} ->
                sample_idx = if use_sample, do: 0, else: 1

                # Check if result matches expected
                expected_idx = (part - 1) * 2 + sample_idx
                expected = Enum.at(expected_results, expected_idx)
                new_match = match and result == expected

                IO.puts(
                  "P#{part} #{Enum.at(@sample_str, sample_idx)} in #{:io_lib.format("~10.4f", [elapsed])}s:    #{ProblemResult.format(result)}"
                )

                {time_acc + elapsed, new_match}

              {:error, reason} ->
                sample_idx = if use_sample, do: 1, else: 0
                IO.puts("P#{part} #{Enum.at(@sample_str, sample_idx)}:  <Error: #{reason}>")
                {time_acc, match}
            end
        end

      IO.puts("\nTotal Runtime: #{:io_lib.format("~.4f", [day_elapsed])}s")
      IO.puts("Note: Results #{if matches_expected, do: "", else: "don't "}match expected")
    end
  end

  defp run_single(runner, day, part, use_sample) do
    case get_result(runner, day, part, use_sample) do
      {:ok, {result, elapsed}} ->
        sample_str = if use_sample, do: "samp", else: "real"

        IO.puts(
          "Day #{String.pad_leading(Integer.to_string(day), 2, "0")} / part #{part} / Data '#{sample_str}' => #{:io_lib.format("~.4f", [elapsed])}s"
        )

        IO.puts(ProblemResult.format(result))

      {:error, reason} ->
        IO.puts("Error: #{reason}")
    end
  end

  defp get_input(runner, day, part, use_sample) do
    base_filename = if use_sample, do: "sample", else: "input"
    day_str = String.pad_leading(Integer.to_string(day), 2, "0")
    filename = "#{runner.year}/inputs/#{base_filename}#{day_str}.txt"

    case ProblemInput.read(filename) do
      {:ok, input} ->
        {:ok, input}

      {:error, :enoent} ->
        # Try versioned filename
        versioned_filename = "#{runner.year}/inputs/#{base_filename}#{day_str}_#{part}.txt"

        case ProblemInput.read(versioned_filename) do
          {:ok, input} -> {:ok, input}
          {:error, _} -> {:error, :no_input}
        end

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp get_result(runner, day, part, use_sample) do
    solution = Enum.at(runner.solutions, day - 1)

    if solution == nil do
      {:error, :missing_solution}
    else
      case get_input(runner, day, part, use_sample) do
        {:ok, input} ->
          {result, duration} = Solution.solve(solution, input, part, use_sample)
          {:ok, {result, duration}}

        {:error, :no_input} ->
          {:ok, {:no_input, 0.0}}

        {:error, reason} ->
          {:error, reason}
      end
    end
  end
end

File: lib/python/runner.py
"""Advent of Code runner with CLI parsing."""

import argparse
import os
import sys
import traceback
from typing import Any

from .solution import ProblemInput, ProblemResult, Solution


class Runner:
    """Runner for executing Advent of Code solutions."""

    def __init__(self, year: int, solutions: dict[int, type[Solution]]):
        """
        Creates a new runner for a given year with solution classes.

        Args:
            year: The year of the Advent of Code
            solutions: Dictionary mapping day numbers (1-25) to Solution classes
        """
        self.year = year
        self.solutions = solutions

    @staticmethod
    def parse_args(args: list[str] | None = None) -> tuple[str | tuple[str, int], dict[str, Any]]:
        """
        Parses command line arguments and returns configuration.

        Returns:
            Tuple of (command, options_dict) where options contains:
            - all: bool
            - part: int (1 or 2)
            - use_sample: bool
        """
        parser = argparse.ArgumentParser(description="Advent of Code - Python Solutions")
        subparsers = parser.add_subparsers(dest="command", help="Command to run")

        # Main command
        subparsers.add_parser("main", help="Run all implemented days")

        # Day command
        day_parser = subparsers.add_parser("day", help="Run a specific day")
        day_parser.add_argument("day", type=int, help="Day number (1-25)")
        day_parser.add_argument("-s", "--sample", action="store_true", help="Use sample input")
        day_parser.add_argument(
            "-p",
            "--part",
            type=int,
            choices=[1, 2],
            help="Run only part 1 or part 2 (required unless --all)",
        )
        day_parser.add_argument(
            "-a",
            "--all",
            action="store_true",
            help="Run both parts (overrides --part)",
        )

        parsed = parser.parse_args(args)

        if parsed.command == "main":
            return "main", {"all": True, "part": 1, "use_sample": False}

        elif parsed.command == "day":
            day = parsed.day
            if not (1 <= day <= 25):
                print("Error: Day must be between 1 and 25")
                sys.exit(1)

            if not parsed.all and parsed.part is None:
                print("Error: --part <1|2> is required unless --all is given")
                sys.exit(1)

            options = {
                "all": parsed.all if parsed.all else False,
                "part": parsed.part if parsed.part else 1,
                "use_sample": parsed.sample if parsed.sample else False,
            }
            return ("day", day), options

        else:
            parser.print_help()
            sys.exit(1)

    def run(
        self, command: str | tuple[str, int], full_day: bool, part: int, use_sample: bool
    ) -> None:
        """
        Runs the solutions based on parsed configuration.

        Args:
            command: Command string ("main" or "dayXX")
            full_day: Whether to run all parts
            part: Part number (1 or 2)
            use_sample: Whether to use sample input
        """
        if command is None:
            print("Usage: python -m <year>.python.main <command>")
            return

        if command == "main":
            self._run_full_year()
        elif command[0] == "day":
            try:
                day = int(command[1])
                if full_day:
                    self._run_day(day)
                else:
                    self._run_single(day, part, use_sample)
            except ValueError:
                print(f"Invalid command format: {command}")
        else:
            print(f"Unknown command: {command}")

    def verify_solution(self, day: int, part: int, use_sample: bool) -> tuple[bool, str | None]:
        """
        Verifies that a solution matches the expected result.

        Returns:
            Tuple of (success, error_message)
        """
        if day not in self.solutions:
            return False, f"No solution implemented for day {day}"

        solution_class = self.solutions[day]
        expected_index = (part - 1) * 2 + (0 if use_sample else 1)
        expected = solution_class.results()[expected_index]

        try:
            result, _ = self._get_result(day, part, use_sample)
            if result == expected:
                return True, None
            else:
                got = self._format_result(result)
                exp = self._format_result(expected)
                return False, f"Got {got}, expected {exp}"
        except Exception as e:
            return False, f"Error: {str(e)}"

    def verify_solutions(self) -> bool:
        """
        Verifies all solutions against expected results.

        Returns:
            True if all tests pass
        """
        print("\n----------")
        all_success = True

        for day in sorted(self.solutions.keys()):
            for part in [1, 2]:
                for use_sample in [True, False]:
                    sample_label = "s" if use_sample else "r"
                    success, error = self.verify_solution(day, part, use_sample)

                    print(f"Testing D{day:02d} P{part} '{sample_label}': ", end="")

                    if success:
                        print("PASSED")
                    else:
                        print(f"FAILED: {error}")
                        all_success = False

        print("----------")
        return all_success

    def _get_input(self, day: int, part: int, use_sample: bool) -> ProblemInput:
        """Load input file for the problem."""
        base_filename = "sample" if use_sample else "input"
        filename = f"{self.year}/inputs/{base_filename}{day:02d}.txt"

        if not os.path.exists(filename):
            # Try versioned filename
            filename = f"{self.year}/inputs/{base_filename}{day:02d}_{part}.txt"

        if not os.path.exists(filename):
            raise FileNotFoundError(f"Input file not found: {filename}")

        with open(filename, encoding="utf-8") as f:
            content = f.read().strip()  # Remove leading/trailing whitespace
            lines = [line.rstrip("\r") for line in content.split("\n")]
            return ProblemInput(lines)

    def _get_result(self, day: int, part: int, use_sample: bool) -> tuple[ProblemResult, float]:
        """
        Get the result for a specific day/part/sample combination.

        Returns:
            Tuple of (result, duration_in_seconds)
        """
        if day not in self.solutions:
            raise ValueError(f"No solution implemented for day {day}")

        solution_class = self.solutions[day]
        input_data = self._get_input(day, part, use_sample)

        # Use Solution.solve static method
        return Solution.solve(solution_class, input_data, part, use_sample)

    @staticmethod
    def _format_result(result: ProblemResult) -> str:
        """Format a result for display."""
        if result is None:
            return "<No Solution>"
        return str(result)

    def _run_full_year(self) -> None:
        """Run all implemented days."""
        print()
        total_time = 0.0

        for day in sorted(self.solutions.keys()):
            print(f"Day {day:02d}:")
            day_time = 0.0

            for part in [1, 2]:
                for use_sample in [True, False]:
                    try:
                        result, duration = self._get_result(day, part, use_sample)
                        day_time += duration
                        sample_str = "samp" if use_sample else "real"
                        print(f"  P{part} {sample_str}:  {self._format_result(result)}")
                    except FileNotFoundError:
                        sample_str = "samp" if use_sample else "real"
                        print(f"  P{part} {sample_str}:  <Error>")

            print(f"  > Runtime: {day_time:.4f}s\n")
            total_time += day_time

        print(f"\n\nTotal Runtime: {total_time:.4f}s")

    def _run_day(self, day: int) -> None:
        """Run all parts of a specific day."""
        if day not in self.solutions:
            print(f"No solution implemented for day {day:02d} in year {self.year}")
            return

        solution_class = self.solutions[day]
        expected_results = solution_class.results()

        day_elapsed = 0.0
        matches_expected = True

        for part in [1, 2]:
            for use_sample in [True, False]:
                sample_str = "samp" if use_sample else "real"

                try:
                    result, duration = self._get_result(day, part, use_sample)
                    day_elapsed += duration

                    sample_idx = 0 if use_sample else 1
                    expected_idx = (part - 1) * 2 + sample_idx
                    expected = expected_results[expected_idx]
                    matches_expected = matches_expected and (result == expected)

                    formatted = self._format_result(result)
                    print(f"P{part} {sample_str} in {duration:10.4f}s:    {formatted}")
                except Exception as e:
                    print(f"P{part} {sample_str}:  <Error: {str(e)}>")

        print(f"\nTotal Runtime: {day_elapsed:.4f}s")
        match_str = "match" if matches_expected else "don't match"
        print(f"Note: Results {match_str} expected")

    def _run_single(self, day: int, part: int, use_sample: bool) -> None:
        """Run a single part of a specific day."""
        try:
            result, elapsed = self._get_result(day, part, use_sample)
            sample_str = "samp" if use_sample else "real"

            print(f"Day {day:02d} / part {part} / Data '{sample_str}' => {elapsed:.4f}s")
            print(self._format_result(result))
        except Exception:
            traceback.print_exc()
